test/functorch/test_control_flow.py

```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+        if compile_mode == "compile":
+            fct_cmp = torch.compile(
+                chain_fct_different_dim, fullgraph=True, dynamic=False
+            )
+        elif compile_mode == "compile_dynamic_shape":
+            fct_cmp = torch.compile(
+                chain_fct_different_dim, fullgraph=True, dynamic=True
+            )
+        elif compile_mode == "eager":
+            fct_cmp = torch.compile(
+                chain_fct_different_dim, fullgraph=True, backend="eager"
+            )
+        else:
+            fct_cmp = chain_fct_different_dim
+
+        expected_result = _fake_associative_scan(
+            add,
+            _fake_associative_scan(add, inp, 1, reverse=reverse),
+            0,
+            reverse=reverse,
         )
-        expected_result = _fake_associative_scan(fct_pointwise, inp, 0, reverse=reverse)
+        result1 = fct_cmp(inp)
         self.assertEqual(result1, expected_result)
-        self.assertEqual(result2, expected_result)
 
     @unittest.skipIf(not SM70OrLater, "triton")
-    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @requires_cuda
+    @parametrize("compile_mode", ["none", "eager", "compile", "compile_dynamic_shape"])
     @parametrize("reverse", [False, True])
     @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
-    def test_generic_associative_scan_generic_simple(self, reverse, device):
+    # Skipping the scan tests, as scan currently does not support inductor
+    @decorateIf(
+        unittest.skip,
+        lambda params: params["compile_mode"] != "eager"
+        and params["compile_mode"] != "none",
+    )
+    def test_scan_downstream_op(self, compile_mode, reverse, device):
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y, x + y
+
+        inp = torch.randn(3, 10, 2, device=device)
+        init = torch.randn(3, 1, 2, device=device)
+
+        for ind in range(2):
+            # Chain with matmul
+            def chain_fct(inp):
+                W = torch.ones(2, 5, device=device)
+                o = scan(add, init=init, input=inp, dim=1, reverse=reverse)
+                return o[ind] @ W
+
+            torch.compiler.reset()
+            if compile_mode == "compile":
+                fct_cmp = torch.compile(chain_fct, fullgraph=True, dynamic=False)
+            elif compile_mode == "compile_dynamic_shape":
+                fct_cmp = torch.compile(chain_fct, fullgraph=True, dynamic=True)
+            elif compile_mode == "eager":
+                fct_cmp = torch.compile(chain_fct, fullgraph=True, backend="eager")
+            else:
+                fct_cmp = chain_fct
+
+            expected_result = _fake_scan(
+                add, init=init, input=inp, dim=1, reverse=reverse
+            )[ind] @ torch.ones(2, 5, device=device)
+            result1 = fct_cmp(inp)
+            self.assertEqual(result1, expected_result)
+
+        # Chain with scan
+        def chain_fct_same_dim(inp):
```

Comment:
Maybe worth separate tests for chained tests.

Reply:
Done